### ISSUE TITLE
Create Weighbridge Ticket doctype

### ISSUE THREAD

**Issue by gaurav-naik** [gaurav-naik/andesit_karang_anyar#23]
Use test.andesitkaranganyar for reference.

Use separate link fields for Customer and Supplier.

**Comment by saidsl** [comment]
A few of Items

1. Would it be possible to increase the Font Size for the Weighings only
2. Please show Driver Name instead of License Number on Screen as well as Print
3. It probably has not been added but I don't see the Item/Product on the WBT Yet

**Comment by gaurav-naik** [comment]
1. Font attributes are not open for direct customization. Will try using jQuery to achieve this.

2. Will make sure that Driver Name is seen in both list and print.
3. Items table:

    Suggested alternative: Link field for items, with text field/child table for charges.
    We see the following issues with having a child table for items:
    1. One item per WBT: If this is true, having a link field for items along with a text field for charges would be more economical. A child table could be added for charges in case there are multiple charges.
    2. UOM is already being captured in the weighing section. It will have to be entered again in the items table. Also, can we have a single UOM field instead of a separate UOM for each weighing?
    3. Qty is being captured as first and second weighing. It will have to be entered again in the items table.

4. Driver to Vehicle would be a many-to-many relationship which can be implemented with a new child doctype under Vehicle. Assigning drivers to a vehicle would become an additional data-entry step. The link field for Driver, currently on the WBT form, allows a partial-matching lookup on Licence No as well as Driver Name. Wouldn't that be sufficient?

**Comment by saidsl** [comment]
Hi Guys — Following are my comments

If Font attributes are not open, that’s fine. Just want to make the numbers more visible
Great
Items Table
It will always be one item on the Supply side and also one Item on the Customer side with the addition of charges (Delivery, service fee, etc)
UOM will be set to default for the item anyway so that should not be an issue
QTY must not be entered manually in ITEMs but obtained/Captured from the NETT Weight result of both Incoming & Outgoing Loads
As I mentioned, I am only concerned that if the list grows to long it will not be User Friendly when selecting the driver. I don’t mind adding an extra step for assigning Drivers to Vehicles as long as its done while entering the Vehicle Details and not a separately as in done on a separate window/screen

Regards
Said


On 28 Jun 2016, at 2:14 PM, gaurav-naik <notifications@github.com> wrote:

Font attributes are not open for direct customization. Will try using jQuery to achieve this.

Will make sure that Driver Name is seen in both list and print.

Items table:

Suggested alternative: Link field for items, with text field/child table for charges.
We see the following issues with having a child table for items:

One item per WBT: If this is true, having a link field for items along with a text field for charges would be more economical. A child table could be added for charges in case there are multiple charges.
UOM is already being captured in the weighing section. It will have to be entered again in the items table. Also, can we have a single UOM field instead of a separate UOM for each weighing?
Qty is being captured as first and second weighing. It will have to be entered again in the items table.
Driver to Vehicle would be a many-to-many relationship which can be implemented with a new child doctype under Vehicle. Assigning drivers to a vehicle would become an additional data-entry step. The link field for Driver, currently on the WBT form, allows a partial-matching lookup on Licence No as well as Driver Name. Wouldn't that be sufficient?

**Comment by saidsl** [comment]
I've just tested the WBT and there are a few things that need to be changed
1. Upon approving WBT (For Customer), The "Create Sales Doc" Button is active, instead can I have a dropdown instead that allow the operator to select whether a Sales Doc or a Delivery Note is created.
2. The same would apply to the WBT (For Supplier), in the dropdown the operator would have the option to create the Purchase Order, Purchase Receipt of Purchase Invoice.

In the case of the Purchase Receipt & Delivery Note, I understand that the respective PO & SO be created in the background, as is actually already done now.

I would like to avoid the case where the operator would have to go to the SO or PO to complete the process. Ideally the entire workflow is done while at the WBT Doc

Thanks

**Comment by revant** [comment]
@saidsl
Please verify Weighbridge Ticket Workflow.

**Comment by saidsl** [comment]
Hi Revant

I will email you tomorrow morning my comments re WBT workflow

Thanks & Regards
Said


On 1 Jul 2016, at 7:06 PM, Revant Nandgaonkar <notifications@github.com> wrote:

@saidsl <https://github.com/saidsl>
Please verify Weighbridge Ticket Workflow.

**Comment by saidsl** [comment]
The WBT workflow is fine except for a few things

1. We need to prevent duplication of Purchase Receipts and Delivery Notes and this duplication happens because when a WBT ticket generates a Purchase Receipt it also generates a Purchase Order when then would allow a Purchase Receipt be created based on that Purchase Order.
2. Since there are instances where a PO is not required and only a PR is, we may need to provide a way where only the PR is created without any PO.
3. There are also instances where multiple PRs are generated from a single PO for large quantities.
4. The same applies to SOs as well, duplication is happening on the DN side.

Basically, other than the duplication issue, we are pretty much there already.

**Comment by revant** [comment]
https://github.com/gaurav-naik/andesit_karang_anyar/commit/c78c3e2c779fc1cdc629c2cf8901e2208fe0b08c

@saidsl please verify and comment.

**Comment by saidsl** [comment]
Hi Guys,

Following are my comments

The workflow works fine, however, the price is still appearing on the Delivery Note Print Version, please refer to my last comments and had requested that price should not be included on the Delivery Note Printed version. On screen it does not matter.
The Driver to Vehicle issue is still not resolved, in that I am still getting the complete list of Drivers.

Other than the above 2 points all is good.

Regards
Said




On 3 Jul 2016, at 8:29 PM, Revant Nandgaonkar <notifications@github.com> wrote:

c78c3e2 <https://github.com/gaurav-naik/andesit_karang_anyar/commit/c78c3e2c779fc1cdc629c2cf8901e2208fe0b08c>
@saidsl <https://github.com/saidsl> please verify and comment.

**Comment by gaurav-naik** [comment]
@saidsl 
Hi.
All fields related to Amount have now been hidden in Delivery Note print. In the previous commit, only the rate field had been hidden.

**Comment by saidsl** [comment]
@gaurav-naik 

Hi, my apologies, I should have mentioned that the price on the Purchase Receipt Print should also not be present.

Is it possible to have this as an optional feature by way of using the Print Format? for both PR & DN

Thanks

**Comment by gaurav-naik** [comment]
@saidsl 

Yes. Fields related to price can be hidden/shown via a checkbox in Delivery Note doctype labelled 'Print Without Amount'. This is a default feature which I discovered after looking through the doctype. I will undo the changes to the price fields from the previous commit, so that the default feature can work correctly.

Purchase Receipt does not have this feature built-in. I will try and duplicate it there. In case this approach cannot be implemented, the Print Format approach can be considered.

Is this ok?

**Comment by saidsl** [comment]
@gaurav-naik 

Yes that's ok, I agree with the approach you are suggesting.

Thanks
Said

**Comment by gaurav-naik** [comment]
@saidsl 

Hi Said.

Now that licence no. is no longer a mandatory field, should the driver dropdown in Weighbridge Ticket now show only those drivers which have a valid licence no?

**Comment by saidsl** [comment]
Hi Gaurav,


Its fine as it is, we’re currently doing a live run and will update soon on a few things

Regards
Said



On 1 Aug 2016, at 3:20 PM, gaurav-naik <notifications@github.com> wrote:

@saidsl <https://github.com/saidsl>
Hi Said.

Now that licence no. is no longer a mandatory field, should the driver dropdown in Weighbridge Ticket now show only those drivers which have a valid licence no?

**Comment by saidsl** [comment]
Hi Gaurav, 

Can you please check on your end if you had ensure that duplication validation has been enabled for all docs from the WBT.

For example, It seems I am getting duplicate Delivery Notes for the same Sales Invoices somehow. 

Can you please ensure that once a document has been completed/submited/processed from the WBT that its no longer available. Currently they all still are even though they have been made

Thanks & Regards
Said




On 1 Aug 2016, at 3:20 PM, gaurav-naik <notifications@github.com> wrote:

@saidsl <https://github.com/saidsl>
Hi Said.

Now that licence no. is no longer a mandatory field, should the driver dropdown in Weighbridge Ticket now show only those drivers which have a valid licence no?

**Comment by gaurav-naik** [comment]
Hi Said.

I will look into the issue in an hour's time.

Thanks.

**Comment by saidsl** [comment]
And couple of things, can we show what process each WBT is at so that its clear what process is currently being done.

And in the meantime, till we finally connect to the Weighbridge Console, can we make the Time In  editable

Thanks & Regards
Said



On 8 Aug 2016, at 10:57 AM, Said Salameh <ss@saidsalameh.com> wrote:

Hi Gaurav, 

Can you please check on your end if you had ensure that duplication validation has been enabled for all docs from the WBT.

For example, It seems I am getting duplicate Delivery Notes for the same Sales Invoices somehow. 

Can you please ensure that once a document has been completed/submited/processed from the WBT that its no longer available. Currently they all still are even though they have been made

Thanks & Regards
Said




On 1 Aug 2016, at 3:20 PM, gaurav-naik <notifications@github.com <mailto:notifications@github.com>> wrote:

@saidsl <https://github.com/saidsl>
Hi Said.

Now that licence no. is no longer a mandatory field, should the driver dropdown in Weighbridge Ticket now show only those drivers which have a valid licence no?

**Comment by saidsl** [comment]
Hi Gaurav. 

Please create a backup before updating anything to the live server and please advice me prior to doing the updates. I will do back ups on my end as well. 

Regards
Said

**Comment by gaurav-naik** [comment]
Hi Said.

Can you share a specific case of duplication of delivery notes?

**Comment by gaurav-naik** [comment]
@saidsl 
Hi Said.

Only one delivery note can be created against one weighbridge ticket. This duplication check has been implemented within Weighbridge ticket. (Clicking on Make > Delivery Note will flash a message saying that one already exists against the ticket.)

However, there is no restriction that would prevent the user from creating multiple delivery notes from a sales order which may have been made against a Weighbridge Ticket.

Would you like us to add such a restriction for Delivery Note and other documents as well?

For any document in ERPNext, multiple instances of the same document can exist in Draft form, but if one of them is submitted, the rest become invalid and cannot be submitted. The restriction would override this flow. Just letting you know.

**Comment by saidsl** [comment]
Hi Gaurav,

Please apply this for the WBT only and leave the standard duplication validation/checks/flow in place.

Thanks & Regards
Said




On 11 Aug 2016, at 2:27 PM, gaurav-naik <notifications@github.com> wrote:

@saidsl <https://github.com/saidsl> 
Hi Said.

Only one delivery note can be created against one weighbridge ticket. This duplication check has been implemented within Weighbridge ticket. (Clicking on Make > Delivery Note will flash a message saying that one already exists against the ticket.)

However, there is no restriction that would prevent the user from creating multiple delivery notes from a sales order which may have been made against a Weighbridge Ticket.

Would you like us to add such a restriction for Delivery Note and other documents as well?

For any document in ERPNext, multiple instances of the same document can exist in Draft form, but if one of them is submitted, the rest become invalid and cannot be submitted. The restriction would override this flow. Just letting you know.

**Comment by gaurav-naik** [comment]
@saidsl 

This validation already exists in WBT.

For instance, if a Delivery Note has already been created against a WBT, clicking on 
Make > Delivery Note in WBT again will display a message indicating that one already exists.

So, just to clarify, we won't be adding any duplication restrictions to other documents such as Sales Order. Is this correct?

**Comment by saidsl** [comment]
Hi Gaurav, correct, if it already exits there’s no need too.

Regards
Said


On 11 Aug 2016, at 3:14 PM, gaurav-naik <notifications@github.com> wrote:

@saidsl <https://github.com/saidsl>
This validation already exists in WBT.

For instance, if a Delivery Note has already been created against a WBT, clicking on 
Make > Delivery Note in WBT again will display a message indicating that one already exists.

So, just to clarify, we won't be adding any duplication restrictions to other documents such as Sales Order. Is this correct?

**Comment by saidsl** [comment]
Hi Gaurav,

Can you please remind/refresh my memory why a WB Settings was done and it only contains the warehouse. At the moment all incoming and outgoing are going out of raw material and I can not add another warehouse as its asking for a different company. How can we change that.

Thanks & Regards
Said





On 11 Aug 2016, at 4:40 PM, Said Salameh <ss@saidsalameh.com> wrote:

Hi Gaurav, correct, if it already exits there’s no need too.

Regards
Said


On 11 Aug 2016, at 3:14 PM, gaurav-naik <notifications@github.com <mailto:notifications@github.com>> wrote:

@saidsl <https://github.com/saidsl>
This validation already exists in WBT.

For instance, if a Delivery Note has already been created against a WBT, clicking on 
Make > Delivery Note in WBT again will display a message indicating that one already exists.

So, just to clarify, we won't be adding any duplication restrictions to other documents such as Sales Order. Is this correct?

**Comment by gaurav-naik** [comment]
Hi Said,

Settings page is a way to set defaults and constants and make them available anywhere in the app. For WB Management, the only default required so far is the Warehouse, which is set in mandatory fields in line-items of documents like Purchase Order and Delivery Note while making them from WBT.

The idea was to have only one default warehouse, so a validation was added that allows only one warehouse for a Company. Technically this is correct.

However, to solve the the current problem, a warehouse field can be added to the WBT doctype. That way, the warehouse can be changed as required and while making documents from WBT, the warehouse would be picked up from WBT instead of picking it up directly from Settings .

Would this be suitable?

**Comment by saidsl** [comment]
HI Gaurav,

Yes that would be more suitable since there is more than one warehouse.

Regards
Said


On 13 Aug 2016, at 12:12 AM, gaurav-naik <notifications@github.com> wrote:

Hi Said,

Settings page is a way to set defaults and constants and make them available anywhere in the app. For WB Management, the only default required so far is the Warehouse, which is set in mandatory fields in line-items of documents like Purchase Order and Delivery Note while making them from WBT.

The idea was to have only one default warehouse, so a validation was added that allows only one warehouse for a Company. Technically this is correct.

However, to solve the the current problem, a warehouse field can be added to the WBT doctype. That way, the warehouse can be changed as required and while making documents from WBT, the warehouse would be picked up from WBT instead of picking it up directly from Settings .

Would this be suitable?

**Comment by saidsl** [comment]
Hi Gaurav

With regards to the Warehouse issue, It would be a far better solution to use the item selected as the basis to default to the correct warehouse instead of selecting the warehouse in the WBT.

Please lets discuss before proceeding with any changes.

Regards
Said


On 13 Aug 2016, at 8:18 AM, Said Salameh <ss@saidsalameh.com> wrote:

HI Gaurav,

Yes that would be more suitable since there is more than one warehouse.

Regards
Said


On 13 Aug 2016, at 12:12 AM, gaurav-naik <notifications@github.com <mailto:notifications@github.com>> wrote:

Hi Said,

Settings page is a way to set defaults and constants and make them available anywhere in the app. For WB Management, the only default required so far is the Warehouse, which is set in mandatory fields in line-items of documents like Purchase Order and Delivery Note while making them from WBT.

The idea was to have only one default warehouse, so a validation was added that allows only one warehouse for a Company. Technically this is correct.

However, to solve the the current problem, a warehouse field can be added to the WBT doctype. That way, the warehouse can be changed as required and while making documents from WBT, the warehouse would be picked up from WBT instead of picking it up directly from Settings .

Would this be suitable?

**Comment by gaurav-naik** [comment]
Hi Said,

The Default Warehouse change you requested has been made, but hasn't been released on the live server yet because of a doubt I needed to clear:

Under Buying Settings, have Purchase Receipts been made mandatory for Purchase Invoices? 
```
(Buying Settings > Purchase Receipt Required > Yes)
```
If so, then a change will have to be made where the Purchase Receipt created against a WBT will be picked up and set in the Purchase Invoice being created against that WBT.

@revant

**Comment by saidsl** [comment]
Hi Gaurav,

Yes, Purchase Receipts been made mandatory for Purchase Invoices.





Regards
Said


On 29 Aug 2016, at 11:08 AM, gaurav-naik <notifications@github.com> wrote:

Hi Said,

The Default Warehouse change you requested has been made, but hasn't been released on the live server yet because of a doubt I needed to clear:

Under Buying Settings, have Purchase Receipts been made mandatory for Purchase Invoices?

(Buying Settings > Purchase Receipt Required > Yes)
If so, then a change will have to be made where the Purchase Receipt created against a WBT will be picked up and set in the Purchase Invoice being created against that WBT.

@revant <https://github.com/revant>

**Comment by gaurav-naik** [comment]
In that case I will make the change I mentioned and update the release later today.

**Comment by saidsl** [comment]
Ok Great, but please don’t push the change to the live server before 5pm Indonesia Time. Or if you send me the instructions I will do so later this evening from my end.

Thanks
Said




On 29 Aug 2016, at 11:16 AM, gaurav-naik <notifications@github.com> wrote:

In that case I will make the change I mentioned and update the release later today.

**Comment by gaurav-naik** [comment]
Ok. I will only push the changes on GitHub after I am done. I will also see about the instructions for pushing changes to the live server.

Thanks.
